src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1019 @@
+  GRPC_ARES_WRAPPER_TRACE_LOG("OnWritable: fd: %d; request:%p; status: %s",
+                              fd_node->as, this, status.ToString().c_str());
+  if (status.ok() && !shutting_down_) {
+    ares_process_fd(channel_, ARES_SOCKET_BAD, fd_node->as);
+  } else {
+    // If error is not absl::OkStatus() or the resolution was cancelled, it
+    // means the fd has been shutdown or timed out. The pending lookups made
+    // on this request will be cancelled by the following ares_cancel(). The
+    // remaining file descriptors in this request will be cleaned up in the
+    // following Work() method.
+    ares_cancel(channel_);
+  }
+  Work();
+}
+
+void GrpcAresRequest::OnQueryTimeout() {
+  grpc_core::MutexLock lock(&mu_);
+  query_timeout_handle_.reset();
+  GRPC_ARES_WRAPPER_TRACE_LOG("request:%p OnQueryTimeout. shutting_down_=%d",
+                              this, shutting_down_);
+  if (!shutting_down_) {
+    shutting_down_ = true;
+    ShutdownPolledFdsLocked(
+        grpc_core::StatusCreate(absl::StatusCode::kDeadlineExceeded,
+                                "OnQueryTimeout", DEBUG_LOCATION, {}));
+  }
+}
+
+// In case of non-responsive DNS servers, dropped packets, etc., c-ares has
+// intelligent timeout and retry logic, which we can take advantage of by
+// polling ares_process_fd on time intervals. Overall, the c-ares library is
+// meant to be called into and given a chance to proceed name resolution:
+//   a) when fd events happen
+//   b) when some time has passed without fd events having happened
+// For the latter, we use this backup poller. Also see
+// https://github.com/grpc/grpc/pull/17688 description for more details.
+void GrpcAresRequest::OnAresBackupPollAlarm() {
+  grpc_core::MutexLock lock(&mu_);
+  ares_backup_poll_alarm_handle_.reset();
+  GRPC_ARES_WRAPPER_TRACE_LOG(
+      "request:%p OnAresBackupPollAlarm shutting_down=%d.", this,
+      shutting_down_);
+  if (!shutting_down_) {
+    for (auto it = fd_node_list_->begin(); it != fd_node_list_->end(); it++) {
+      if (!(*it)->already_shutdown) {
+        GRPC_ARES_WRAPPER_TRACE_LOG(
+            "request:%p OnAresBackupPollAlarm; ares_process_fd. fd=%s", this,
+            (*it)->polled_fd->GetName());
+        ares_socket_t as = (*it)->polled_fd->GetWrappedAresSocketLocked();
+        ares_process_fd(channel_, as, as);
+      }
+    }
+    if (!shutting_down_) {
+      EventEngine::Duration next_ares_backup_poll_alarm_duration =
+          calculate_next_ares_backup_poll_alarm_duration();
+      ares_backup_poll_alarm_handle_ = event_engine_->RunAfter(
+          next_ares_backup_poll_alarm_duration,
+          [self = Ref(DEBUG_LOCATION, "OnAresBackupPollAlarm")]() mutable {
+            grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
+            grpc_core::ExecCtx exec_ctx;
+            self->OnAresBackupPollAlarm();
+          });
+    }
+    Work();
+  }
+}
+
+// TODO(yijiem): Consider report this status or as part of the result when
```

Comment:
There was some issues with the error handling in this code. I made changes so that:
1. If there is valid result, always return that through the `absl::StatusOr` argument to the `on_resolve` callback (except cancellation).
2. If error needs to be reported, the `absl::Status`'s code will be the first error status code that is encountered, e.g. if timeout happened to be the first exceptional event, the status code would be `kDeadlineExceeded`. And the `absl::Status`'s message will be a composite (concatenated) message as the code hits further error status. This is done through a newly added `AddError` method. PTAL. Thanks!